### PR TITLE
Drop support for documents trashcan, garbage collector, and with the _meta tag

### PR DESCRIPTION
**Pull request by scottinet**
# Description

* Remove the `_meta` tag from notifications and from the documentation
* Remove the documents garbage collector (the documents trashcan does not exist anymore)
* Remove the remaining traces of the documents trashcan

### CODE REVIEW

File: lib/api/core/models/repositories/tokenRepository.js [hunk 1]
```diff
@@ -221,6 +221,7 @@ class TokenRepository extends Repository {
       emptyKeyLength = super.getCacheKey('').length,
       userKey = super.getCacheKey(userId + '#*');
 
+    // @todo ask @scottinet or @bvidis for this
```
Comment (kuzzle) [inline lib/api/core/models/repositories/tokenRepository.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/controllers/adminController.js [hunk 2]
```diff
@@ -84,11 +80,16 @@ class AdminController extends BaseController {
...
 
-    const
-      database = request.input.args.database,
-      cacheEngine = this.kuzzle.services.list[database];
+    const database = request.input.args.database;
+
+    let cacheEngine;
 
-    if (!cacheEngine || typeof cacheEngine.flushdb !== 'function') {
+    // @todo allow only publicCache
```
Comment (kuzzle) [inline lib/api/controllers/adminController.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/core/models/repositories/repository.js [hunk 3]
```diff
@@ -59,19 +60,28 @@ class Repository {
...
-    if (options.databaseEngine === null) {
-      this.databaseEngine = options.databaseEngine;
+    // if set to null, it means that we explicitely dont want a indexEngine for this repository
+    // as null is falsy, without this check, the indexEngine is always set...
+    if (options.indexEngine === null) {
+      this.indexEngine = options.indexEngine;
     }
     else {
-      this.databaseEngine = options.databaseEngine || this.kuzzle.internalEngine;
+      // @todo refactor storageEngine set (it's dirty to pass it with options.indexEngine)
```
Comment (kuzzle) [inline lib/api/core/models/repositories/repository.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/core/storage/indexCache.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    if (collection && ! this.indexes[index].collections.includes(collection)) {
+      this.indexes[index].collections.push(collection);
+
+      modified = true;
+    }
+
+    if (modified) {
+      // @todo modify cluster for scope
```
Comment (kuzzle) [inline lib/api/core/storage/indexCache.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/core/plugins/pluginContext.js [hunk 5]
```diff
@@ -75,23 +76,67 @@ class PluginContext {
...
+        get: () => {
+          return {
+            bootstrap: () => pluginIndexEngine.init(),
+            createCollection: (...args) => pluginIndexEngine.createCollection(...args)
+          };
+        }
+      });
+
+      /**
+       * @todo Use a CollectionEngine class instead repository
```
Comment (kuzzle) [inline lib/api/core/plugins/pluginContext.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/controllers/indexController.js [hunk 6]
```diff
@@ -161,7 +161,8 @@ class IndexController extends BaseController {
    * @returns {Promise<Object>}
    */
   refreshInternal() {
-    return this.kuzzle.internalEngine.refresh()
+    // @todo remove refresh* methods
```
Comment (kuzzle) [inline lib/api/controllers/indexController.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

**Comment by kuzzle** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 8 info

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [adminController.js#L96](https://github.com/kuzzleio/kuzzle/blob/cdf7d36ef29acaf6edde6b326db5f54010ca2f1b/lib/api/controllers/adminController.js#L96): TypeError can be thrown as "cacheEngine" might be null or undefined here. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS2259)
1. ![MINOR][MINOR] [redis.js#](https://github.com/kuzzleio/kuzzle/blob/cdf7d36ef29acaf6edde6b326db5f54010ca2f1b/lib/services/redis.js#): Add a new line at the end of this file. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AMissingNewlineAtEndOfFile)
1. ![INFO][INFO] [collectionController.js#L70](https://github.com/kuzzleio/kuzzle/blob/cdf7d36ef29acaf6edde6b326db5f54010ca2f1b/lib/api/controllers/collectionController.js#L70): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [kuzzle.js#L53](https://github.com/kuzzleio/kuzzle/blob/cdf7d36ef29acaf6edde6b326db5f54010ca2f1b/lib/api/kuzzle.js#L53): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 8 issues
* ![INFO][INFO] 8 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [collectionController.js#L70](https://github.com/kuzzleio/kuzzle/blob/074d56b6e764c6b874ab4ef2220cc70ab5b54535/lib/api/controllers/collectionController.js#L70): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [kuzzle.js#L53](https://github.com/kuzzleio/kuzzle/blob/074d56b6e764c6b874ab4ef2220cc70ab5b54535/lib/api/kuzzle.js#L53): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: doc/2/guides/migrate-from-v1/introduction/index.md [hunk 7]
```diff
@@ -14,7 +14,12 @@ Dropped support for:
   - Redis versions 3 and 4
   - Elasticsearch v5 
   - Kuzzle Proxy 
-  - Permission Closures
+  - Permission Closures (deprecated since Kuzzle 1.4.0)
+  - Document trashcan (deprecated since Kuzzle 1.2.0)
+
+API Changes:
+  - Remove the `_meta` tag from documents and notifications (it was deprecated since Kuzzle 1.3.0)
+  - Fields linked to the documents trashcan have been removed from the dDocuments and notifications metadata : `deletedAt`, `active`
```
Comment (Aschen) [inline doc/2/guides/migrate-from-v1/introduction/index.md]:
```suggestion
  - Fields linked to the documents trashcan have been removed from the documents and notifications metadata : `deletedAt`, `active`
```

**Pull request merged**